core/domain/topic_services.py

```diff
@@ -255,15 +258,29 @@ def apply_change_list(topic_id, change_list):
             topic.
 
     Returns:
-        Topic. The resulting topic domain object.
+        dict, list(int), list(int). A dict with two fields: topic and
+            subtopic_pages containing the updated domain objects of each one,
+            a list of ids of the deleted subtopics and a list of ids of the
+            newly created subtopics.
     """
     topic = get_topic_by_id(topic_id)
+    subtopic_pages = subtopic_page_services.get_all_subtopic_pages_in_topic(
+        topic_id)
+    newly_created_subtopic_ids = []
+    deleted_subtopic_ids = []
+    return_dict = {}
     try:
         for change in change_list:
             if change.cmd == topic_domain.CMD_ADD_SUBTOPIC:
-                topic.add_subtopic(change.id, change.title)
+                topic.add_subtopic(change.subtopic_id, change.title)
+                subtopic_pages.append(
+                    subtopic_page_domain.SubtopicPage.create_default_subtopic_page( #pylint: disable=line-too-long
+                        change.subtopic_id, topic_id)
+                )
+                newly_created_subtopic_ids.append(change.subtopic_id)
             elif change.cmd == topic_domain.CMD_DELETE_SUBTOPIC:
                 topic.delete_subtopic(change.id)
+                deleted_subtopic_ids.append(change.id)
```

Comment:
Actually you'll need to check whether it's in newly_created_subtopic_ids first, right? If so, remove it from there...

On further thought -- there is a design question here. If a creator creates and deletes a subtopic in the same commit, should this actually cause next_subtopic_id to be incremented or a subtopic model to be created for that id? If so, then I guess we'd need to create the subtopic page with the latest content and immediately delete it; if not, perhaps the changelist should be "cleaned" in the frontend before being sent to the backend so that any commits related to ephemeral subtopic pages are removed.

Reply:
I am doing that check before I actually delete the subtopic pages in update_topic_and_subtopic_pages().
Yeah, currently, next_subtopic_id is incremented even if it is created and deleted in the same changelist. How about instead of frontend, in update_topic_and_subtopic_pages() itself, I filter these type of changes out, and then call apply_change_list?

Reply:
Yep, that's fine too. Though you'll need to make sure the frontend knows about the new subtopic IDs and readjusts its internal state accordingly.